Query interface

00vareladavid opened the pull request:
This provides a low-level interface that handles queries about the dependency graph.

00vareladavid commented:
Two parts:
- choose a subset of the graph
    - `active`, `direct`, and `all` (could probably use better names)
    - alternatively, can explicitly give a list of UUIDs
- choose properties to query
    - currently supports [`name` `pinned` `rev` `dev` `source` `stdlib` `version` `url`]

KristofferC commented:
An official interface for asking these type of queries (that we keep stable) would greatly help people working on stuff like the VSCode plugin I think.

fredrikekre commented:
I had also planned to do something similar to this. I was thinking of just creating an interface to return the TOML-parsed `Dict` of Projetc.toml and then you can do whatever you want with that. Not sure what is better, or if they are orthogonal to each other.

00vareladavid commented:
Perhaps orthogonal, this is intended for other packages to more easily hook into `Pkg`. It might be too heavy for people who e.g. want to query the version of their own project.

Its possible, but cumbersome as a human interface: `collect(values(Pkg.info(;active=true, version=true)))[1]["version"]`. A user-friendly interface might also support name-based queries, as opposed to UUID-based.

KristofferC commented:
Putting this as a reference https://doc.rust-lang.org/cargo/reference/external-tools.html#information-about-package-structure.

KristofferC commented:
One idea which might be a bit more future proof is to make a dedicated `struct` that holds the information instead of a `Dict{String, Any}`. That was we can deprecate things and tab completion will work etc.

davidanthoff commented:
So in general I would love to see a stable API. In practice, for VS Code, the issue is that we want to support julia 1.0 going forward, given that it is the long term release. So anything that just appears in say in julia 1.2 or julia 1.3 doesn't help us much... In fact, what really would help us would be a package (not in the stdlib) that hides all the differences between different julia 1.x versions for us.

KristofferC commented:
>In fact, what really would help us would be a package (not in the stdlib) that hides all the differences between different julia 1.x versions for us.

I don't think anyone here really has plan on doing that. Using the "query interface" whenever it gets in a release and "ifdef" for earlier versions would likely be your best option.

00vareladavid commented:
bors try

00vareladavid commented:
bors try

00vareladavid commented:
I feel this is good to merge. Any objections/feedback?

As Kristoffer suggested, the data is returned as a struct instead of a dictionary. I also added documentation and some basic tests. I also remove the old `__installed` API which serves a similar purpose but is based on names, not UUIDs.

Since this is supposed to be a bit more future proof. I went with the conservative approach of requiring users to explicitly specify all the properties that they want to query. The idea is that future properties might be more a lot more expensive to compute (I'm thinking of cases where we have to parse registry info). Feedback on this approach specifically?

I was also conservative with the initial set of properties that users can query. We would definitely add more properties as we understand the kinds of questions that users ask.

oxinabox commented on src/Pkg.jl:
```diff
@@ -264,6 +261,47 @@ const develop = API.develop
+| `developed` | `Bool`          | Whether a package is directly tracking a directory |
+| `pinned`    | `Bool`          | Whether a package is pinned                        |
+| `source`    | `String`        | The directory containing the package's source code |
+
```
Which of these will be `nothing` for stdlibs?

00vareladavid replied:
All fields should "Just Work" for stdlibs, with the exception of `version` which will always be `nothing`.

oxinabox commented on src/Pkg.jl:
```diff
@@ -264,6 +261,47 @@ const develop = API.develop
+    Pkg.info(; keywords...)::Dict{UUID, PackageInfo}
+    Pkg.info(uuid::UUID; keywords...)::Dict{UUID, PackageInfo}
+    Pkg.info(uuids::Vector{UUID}; keywords...)::Dict{UUID, PackageInfo}
+
```
Document what the uuid is the uuid of.
I am pretty sure it is a package UUID.

Also you probably want to include instructions somewhere on how to retrieve a list of UUIDs for a given package name and a registry / locally.
(I feel like that is `pkgid`?)

oxinabox commented on src/Types.jl:
```diff
@@ -1348,4 +1348,16 @@ function write_env(ctx::Context; display_diff=true)
+### PackageInfo
+###
+
+Base.@kwdef mutable struct PackageInfo
```
I kinda wish this was written in a way that lets this be non-mutable

oxinabox commented on src/Operations.jl:
```diff
@@ -64,6 +64,7 @@ end
 
 function load_all_deps!(ctx::Context, pkgs::Vector{PackageSpec}; version::Bool=true)
     for (uuid, entry) in ctx.env.manifest
+        pkgs[uuid] === nothing || continue # dont duplicate packages
```
```suggestion
        pkgs[uuid] === nothing || continue  # don't duplicate packages
```

oxinabox commented on src/Operations.jl:
```diff
@@ -1298,4 +1299,16 @@ function test(ctx::Context, pkgs::Vector{PackageSpec}; coverage=false, test_fn=n
+    source    && (info.source = project_rel_path(ctx, source_path(pkg)))
+    developed && (info.developed = pkg.path !== nothing)
+    return info
+end
```
Actually Package info can be made immutable fairly easily.
I think it would be good to do so, since it would be very confusing if someone mutated it in their code.

```
info = PackageInfo(
    name      = name ? pkg.name : nothing,
    version   = version && pkg.version == VersionSpec() ? pkg.version : nothing,
    pinned    = pinned ? pkg.pinned : nothing,
    source    = source ? project_rel_path(ctx, source_path(pkg)) : nothing,
    developed = developed ? pkg.path !== nothing : nothing,
)
return info
```

oxinabox commented on src/API.jl:
```diff
@@ -18,6 +18,28 @@ preview_info() = printstyled("───── Preview mode ─────\n"; c
+        if ctx.env.pkg === nothing
+            @warn "Active environment is not a project"
+        else
+            pkg = deepcopy(ctx.env.pkg)
```
Why is the `deepcopy` needed?

oxinabox commented:
Sorry I was so slow in getting around to looking at this.

> Since this is supposed to be a bit more future proof. I went with the conservative approach of requiring users to explicitly specify all the properties that they want to query. The idea is that future properties might be more a lot more expensive to compute (I'm thinking of cases where we have to parse registry info). Feedback on this approach specifically?

I am not enthused about it. I don't hate it but...

What if we used properties & closures, so they could be calculated when and if the field was checked ?
Or rather planned to use properties & closures, for any expensive ones that might turn up later.

timholy commented:
In case it helps with API design, other questions I *know* we're going to want to ask are:
- "which other registered packages depend on X?"
- "what new incompatibilities are generated if I release version `n` of X"?

Ref https://github.com/JuliaRegistries/Registrator.jl/issues/122. My main concern is that this API seems to focus on "what does X depend on?" whereas I tend to think it's more useful to ask "what depends on X?" Obviously the first is more of a finite set, the latter depends on which registry(ies) you check and there is no guarantee that you know about all the registries that exist in the universe.

00vareladavid commented:
Thanks @oxinabox for the excellent feedback. I agree lazily loading any expensive properties is a better API. And there is indeed no reason why the properties shouldn't be immutable.

00vareladavid commented:
bors try

00vareladavid commented:
This has changed a bit from the initial design:
   - I changed the name from `Pkg.info` to `Pkg.dependencies`. I think this makes the purpose more clear. Is "dependencies" too long or should we go with `Pkg.deps`?
   - Simplified the interface by removing all keyword arguments from `Pkg.dependencies`. It will just return the whole set of dependencies (as a `Dict{UUID,PackageInfo}`)
   - added `Pkg.project` function which returns a `ProjectInfo` struct containing info about the active project. @fredrikekre you mentioned a plan similar to this. Do you think this will suffice? I will finish documenting it if it looks good.
   - any bikeshedding on field names?

00vareladavid commented:
bors try

oxinabox commented on src/Pkg.jl:
```diff
@@ -273,6 +270,30 @@ const develop = API.develop
+    Pkg.dependencies()::Dict{UUID, PackageInfo}
+
+Query the dependecy graph.
+The result is a `Dict` that maps a package UUID to a `PackageInfo` struct representing that package.
```
```suggestion
The result is a `Dict` that maps a package UUID to a `PackageInfo` struct representing that the dependency (a package). 
```

oxinabox commented:
> I changed the name from Pkg.info to Pkg.dependencies. I think this makes the purpose more clear. Is "dependencies" too long or should we go with Pkg.deps?

`dependencies` seems good to me.
I would also be ok with `dependency_packageinfos` to be fully explict.

And also wold go for `projectinfo` over `project` even.

StefanKarpinski commented:
Since the section is called `deps` it seems like `Pkg.deps` would be more consistent.

oxinabox commented:
Though it doesn't just list them.it also gets all their `PkgInfo`.

00vareladavid commented:
bors try

KristofferC commented on src/Pkg.jl:
```diff
@@ -273,6 +270,41 @@ const develop = API.develop
 const generate = API.generate
 
+"""
+    Pkg.dependencies()::Dict{UUID, PackageInfo}
```
Should this include things only in the Project, or also in the Manifest or should that be an option?

KristofferC commented on src/Pkg.jl:
```diff
@@ -273,6 +270,41 @@ const develop = API.develop
+const dependencies = API.dependencies
+
+"""
+    Pkg.project()::ProjectInfo
```
I wonder if `ProjectInfo(::UUID)` would be a good constructor? Otherwise it is a bit hard to do something with the `.dependencies` field.

00vareladavid commented:
bors r+

00vareladavid commented:
bors r+

00vareladavid commented:
bors r+
